### ISSUE TITLE
Problem with getting all ever registered domains of namecoin with name_scan RPC

### ISSUE THREAD

**Issue by 1234qwerasdfyxcv** [namecoin/namecoin-core#341]
Hey guys,

i am having a problem with gettin all ever registered domains (d/...) with the RPC 'name_scan'.
When I run it i get only one and two character long domains sorted alphabetically. When I resolute the active domains ("expired": false) with ncdns i get no answer back, which seems like the domain is not active.
Is that a bug or is that intended

If intended, does someone know how I can check for all ever registered domain names correctly?

Greetings

**Comment by domob1812** [comment]
You should be able to list all names (not just domains but anything) using `name_scan`.  What is the exact command you are using?

**Comment by 1234qwerasdfyxcv** [comment]
Thank you for your fast answer.
I used 'namecoin-cli.exe name_scan' and 'namecoin-cli.exe name_scan all', both giving me different outputs.
I want to have all ever registered .bit domains though. Is that possible with a command?

**Comment by domob1812** [comment]
Yes.  Use a command like this:

    namecoin-cli name_scan "d/domob" 5 '{"prefix":"d/","nameEncoding":"utf8"}'

This will give you the next five `d/` names (i.e. domains), starting from `d/domob` in this case.

So what you should do is start with `"d/"` as first argument and call to get e.g. 1000 names.  Then call again with the last name you got as starting name (first argument) to get the next 1000, and so on until you get fewer than 1000 names.

**Comment by 1234qwerasdfyxcv** [comment]
`namecoin-cli name_scan "d/" 1000000` this command works, but i still get names with id/ prefix at the end. Moreover i have many entrys with `"name_error": "invalid data for ascii".
I get that i have to define nameEncoding as utf8, but when i run your command `namecoin-cli name_scan "d/" 1000000 '{"prefix":"d/","nameEncoding":"utf8"}'` i get the error `error: Error parsing JSON:'{prefix:d/,nameEncoding:utf8}'`.
Is it safe to say that if i scan 1000000 domains, i get all d/ prefixed names?

Thank you very much for your help :)

Greetings

**Comment by domob1812** [comment]
With the `prefix` filter, you should not get `id/` names.  And if you are unsure about the number, just try calling again passing in the last name you got as first argument.

It seems that the error you are getting is related to escaping of the third argument by your shell.  What shell are you using?

**Comment by 1234qwerasdfyxcv** [comment]
But isn't that filter done twice then? I mean using `"d/"` after `name_scan` and with` "prefix":"d/"`?

I mean there are 491,489 blocks, so there shouldn't be more than 491,489 domains if i consider correctly or?

Yeah error is related to not accepting the third argument i guess. I'm using Windows 10 command prompt (cmd). What shell are you using? Bash?

**Comment by domob1812** [comment]
The two `d/` strings are different things.  The first argument `"d/"` tells `name_scan` to *start looking from* the name `d/` (i.e. the start).  You can also put e.g. `d/domob` there, then it will start iterating from `d/domob` onwards.  That's how you can do "pagination" if you limit to, say, 1000 names per call.  The second one (with `prefix`) is the real filter.

There can be more names than blocks.  Initially a name cost 50 NMC (the block reward), but that dropped after a small period of time.  But I guess that 1'000'000 should be enough to cover all domains in one call, if that's what you want to do.

Yeah, I'm using `bash`.  I don't really know how to properly escape the string in the Windows command prompt.

**Comment by 1234qwerasdfyxcv** [comment]
Okay with bash its working.

Yeah I want to parse the domains to python, so i can just check if it is less than 100 mil json data entrys.

Other question, do you know why i don't get A records with IP-Adress back when I resolute active domains using dig in combination with ncdns. For domains like 'jamieweb.bit' or 'nf.bit' i get an A record answer back though.

Maybe u have an idea why it does not work for the ones I got from namecoin-cli name_scan command.

**Comment by domob1812** [comment]
For that other question, @JeremyRand is the one to answer. :)

**Comment by 1234qwerasdfyxcv** [comment]
Ok thank you, I will do that.

I have another problem though. With the command `./namecoin-cli name_scan "d/" 1000 '{"prefix":"d/","nameEncoding":"utf8"}'` I get domain names back like d/   or d/'.
That can't be legit and valid names of .bit domains or?

Moreover the 'expires in' value does not change when I reuse the command few seconds later. That should not be like that or?

**Comment by domob1812** [comment]
Yes, you will get all Namecoin names that start with `d/`.  Not all of them will be valid.  I think you have to filter that externally, and don't think `name_scan` has a way to filter for you.  (You can filter by a regexp instead of just `d/` prefix, so maybe you could craft one that does what you want.  I'm not sure about this, though, especially if you take things like IDN into account.)

`expires_in` is based on blocks, so if you redo the command a few seconds later, then it is expected to stay unchanged unless a block came in right around that time.
